File: core/track_constraints.py
"""Backend twin of the frontend constraint kernel (v3.0.1 M1/M2).

Semantic mirror of ``frontend/src/utils/trackConstraints.ts`` -- both
sides are pinned to the same boundary-case table by their tests (SPEC
M0-1: ``frontend/src/utils/trackConstraints.test.ts`` and
``tests/test_track_constraints.py``). Consumed by the project_service
write channels (SPEC M2). Any semantic change must land on BOTH sides in
the same PR, including the constant values below.

Red lines (SPEC M0-3):
- reconcile never mutates the main track (covered ranges are read-only).
- offsets are always rebuilt wholesale (``rebuild_binding_offsets``),
  never maintained incrementally.
"""

from __future__ import annotations

import math
from collections.abc import Sequence
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from core.models import Segment

MIN_SEGMENT_DURATION = 0.1
SNAP_STEP = 0.01

_EPSILON = 1e-6


def _round3(t: float) -> float:
    # Math.round(t * 1000) / 1000 semantics: half-away-from-zero via
    # floor(x + 0.5) -- Python round() is banker's rounding and would
    # diverge from the TS twin on exact .5 ties.
    return math.floor(t * 1000 + 0.5) / 1000


def _check_finite(value: float, name: str) -> None:
    if not math.isfinite(value):
        raise ValueError(f"track_constraints: {name} must be finite, got {value!r}")


def snap_to_step(time: float, step: float = SNAP_STEP) -> float:
    """Snap ``time`` to the nearest multiple of ``step`` (1/step must be exact)."""
    _check_finite(time, "time")
    if step <= 0:
        raise ValueError("track_constraints: step must be positive")
    inv = 1 / step
    # Math.round semantics (half-up), NOT Python round() -- see _round3.
    return math.floor(time * inv + 0.5) / inv


# ------------------------------------------------------------------
# M1-1: neighbor bounds + main-track cue constraint
# ------------------------------------------------------------------


@dataclass(frozen=True)
class NeighborBounds:
    prev_end: float | None = None
    next_start: float | None = None


@dataclass(frozen=True)
class CueConstrainResult:
    ok: bool
    start: float = 0.0
    end: float = 0.0
    reason: str | None = None
    gap: float = 0.0


def get_track_neighbor_bounds(
    segments: Sequence[Segment],
    segment_id: str,
    moved_ids: set[str] | frozenset[str] | None = None,
) -> NeighborBounds:
    """Neighbor bounds of ``segment_id`` on its own track.

    Segments whose id is in ``moved_ids`` are exempt (multi-select /
    linkage drags must not bound each other). Tolerates unsorted input.
    """
    ordered = sorted(segments, key=lambda s: s.start)
    idx = next((i for i, s in enumerate(ordered) if s.id == segment_id), -1)
    if idx == -1:
        return NeighborBounds()
    prev_end: float | None = None
    for i in range(idx - 1, -1, -1):
        if moved_ids and ordered[i].id in moved_ids:
            continue
        prev_end = ordered[i].end
        break
    next_start: float | None = None
    for i in range(idx + 1, len(ordered)):
        if moved_ids and ordered[i].id in moved_ids:
            continue
        next_start = ordered[i].start
        break
    return NeighborBounds(prev_end=prev_end, next_start=next_start)


def constrain_cue_range_to_track(
    start: float,
    end: float,
    bounds: NeighborBounds,
    min_duration: float = MIN_SEGMENT_DURATION,
) -> CueConstrainResult:
    """Main-track "clamp into the neighbor gap" rule (see TS twin for the
    full rule comment; hug-prev -> hug-next -> cap fallback chain)."""
    _check_finite(start, "start")
    _check_finite(end, "end")
    if end < start:
        start, end = end, start
    original_width = end - start

    lo = bounds.prev_end if bounds.prev_end is not None else -math.inf
    hi = bounds.next_start if bounds.next_start is not None else math.inf
    if hi - lo < min_duration - _EPSILON:
        gap = hi - lo if math.isfinite(hi - lo) else 0.0
        return CueConstrainResult(ok=False, reason="gap-too-narrow", gap=gap)

    s = max(start, lo)
    e = min(end, hi)
    if e - s < min_duration - _EPSILON and original_width >= min_duration - _EPSILON:
        # Keep the ORIGINAL width while sliding (see TS twin note: hug-next
        # stays as a defensive branch; hug-prev overflows iff dur > gap).
        dur = original_width
        s = lo
        e = lo + dur
        if e > hi + _EPSILON:
            e = hi
            s = hi - dur
            if s < lo - _EPSILON:
                s = lo
    return CueConstrainResult(ok=True, start=_round3(s), end=_round3(e))


# ------------------------------------------------------------------
# M1-2: extension-track constraints
# ------------------------------------------------------------------


def clamp_extension_range(
    start: float,
    end: float,
    duration: float,
    min_duration: float = MIN_SEGMENT_DURATION,
) -> tuple[float, float]:
    """Global [0, duration] clamp + minimum duration + round3."""
    _check_finite(start, "start")
    _check_finite(end, "end")
    _check_finite(duration, "duration")
    if duration <= 0:
        return (0.0, 0.0)
    if duration <= min_duration:
        return (0.0, _round3(duration))

    s = min(max(0.0, start), duration - min_duration)
    e = min(max(end, min_duration), duration)
    if e - s < min_duration:
        if s + min_duration <= duration:
            e = s + min_duration
        else:
            s = duration - min_duration
            e = duration
    return (_round3(s), _round3(e))


def overlaps_neighbors(
    start: float,
    end: float,
    segments: Sequence[Segment],
    segment_id: str,
    moved_ids: set[str] | frozenset[str] | None = None,
    epsilon: float = _EPSILON,
) -> bool:
    """O(n) overlap probe; touching edges (gap <= epsilon) do NOT count."""
    _check_finite(start, "start")
    _check_finite(end, "end")
    for s in segments:
        if s.id == segment_id:
            continue
        if moved_ids and s.id in moved_ids:
            continue
        if start < s.end - epsilon and end > s.start + epsilon:
            return True
    return False


# ------------------------------------------------------------------
# M1-3: linkage follow + reconcile
# ------------------------------------------------------------------


@dataclass(frozen=True)
class ReconcileCounters:
    squeezed: int = 0
    removed: int = 0
    unbound: int = 0


@dataclass(frozen=True)
class ReconcileResult:
    segments: list[dict]  # surviving geometry: [{"id", "start", "end"}]
    removed_ids: list[str]
    counters: ReconcileCounters


def reconcile_extension_track(
    ext_segments: Sequence[Segment],
    covered: Sequence[tuple[float, float]],
    min_duration: float = MIN_SEGMENT_DURATION,
) -> ReconcileResult:
    """Passive-side resolution after the main track moved (see TS twin).

    ``covered`` is READ-ONLY input -- reconcile never rewrites the main
    track. Binding dissolution is derived by the caller from
    ``removed_ids`` (1:1 model: removed segment == unbound binding).
    """
    squeezed = 0
    removed: list[str] = []
    surviving: list[dict] = []

    for seg in ext_segments:
        gaps: list[tuple[float, float]] = []
        cursor = seg.start
        for (c_start, c_end) in covered:
            if c_end <= cursor + _EPSILON or c_start >= seg.end - _EPSILON:
                continue
            if c_start > cursor + _EPSILON:
                gaps.append((cursor, min(c_start, seg.end)))
            cursor = max(cursor, min(c_end, seg.end))
            if cursor >= seg.end - _EPSILON:
                break
        if cursor < seg.end - _EPSILON:
            gaps.append((cursor, seg.end))

        if not gaps:
            removed.append(seg.id)
            continue

        best = max(gaps, key=lambda g: g[1] - g[0])
        if best[1] - best[0] >= min_duration - _EPSILON:
            if (
                abs(best[0] - seg.start) > _EPSILON
                or abs(best[1] - seg.end) > _EPSILON
            ):
                squeezed += 1
            surviving.append(
                {"id": seg.id, "start": _round3(best[0]), "end": _round3(best[1])}
            )
        else:
            removed.append(seg.id)

    n_removed = len(removed)
    return ReconcileResult(
        segments=surviving,
        removed_ids=removed,
        counters=ReconcileCounters(
            squeezed=squeezed, removed=n_removed, unbound=n_removed
        ),
    )


def sync_bound_extension_for_main(
    main_before: tuple[float, float],
    main_after: tuple[float, float],
    ext: tuple[float, float],
) -> tuple[float, float]:
    """Main -> extension delta follow (move = equal deltas; trims stack)."""
    d_start = main_after[0] - main_before[0]
    d_end = main_after[1] - main_before[1]
    return (ext[0] + d_start, ext[1] + d_end)


def rebuild_binding_offsets(
    main: tuple[float, float],
    ext: tuple[float, float],
) -> dict[str, float]:
    """Wholescale offset rebuild: offset = ext - main, round3. The ONLY
    sanctioned way to produce binding offsets (red line M0-3)."""
    return {
        "start_offset": _round3(ext[0] - main[0]),
        "end_offset": _round3(ext[1] - main[1]),
    }


# ------------------------------------------------------------------
# M1-4: extension -> main reverse constraint (ported, UI not wired 3.0.1)
# ------------------------------------------------------------------


@dataclass(frozen=True)
class BoundPanelEditResult:
    ok: bool
    main_start: float
    main_end: float
    shifted: float


def constrain_bound_extension_panel_edit(
    delta: float,
    main: tuple[float, float],
    bounds: NeighborBounds,
    min_duration: float = MIN_SEGMENT_DURATION,
) -> BoundPanelEditResult:
    _check_finite(delta, "delta")
    proposed = (main[0] + delta, main[1] + delta)
    r = constrain_cue_range_to_track(proposed[0], proposed[1], bounds, min_duration)
    if not r.ok:
        return BoundPanelEditResult(ok=False, main_start=main[0], main_end=main[1], shifted=0.0)
    return BoundPanelEditResult(
        ok=True, main_start=r.start, main_end=r.end, shifted=r.start - main[0]
    )

File: frontend/src/utils/trackConstraints.ts
/**
 * Stacked-timeline constraint kernel (v3.0.1 M1).
 *
 * Pure-function port of MAW's "constrain first, write state second"
 * geometry kernel (see docs/3.0.1/spec-v3.0.1.md M1). Every function here
 * is framework-free: no Vue, no bridge, no component imports -- inputs and
 * outputs are plain serializable data. The backend keeps a semantic mirror
 * in `core/track_constraints.py`; both sides are pinned to the same
 * boundary-case table by tests (M1 acceptance).
 *
 * Red lines (SPEC M0-3):
 * - reconcile never mutates the main track (covered ranges are read-only).
 * - offsets are always rebuilt wholesale (`rebuildBindingOffsets`), never
 *   maintained incrementally.
 */

import type { Segment, TrackBinding } from "@/types/project"

// ------------------------------------------------------------------
// Constants (single source of truth; SegmentBlocksLayer imports these)
// ------------------------------------------------------------------

/** Minimum segment duration in seconds (mirrored in core/track_constraints.py). */
export const MIN_SEGMENT_DURATION = 0.1

/** Snap step in seconds (mirrored in core/track_constraints.py). */
export const SNAP_STEP = 0.01

const EPSILON = 1e-6

function assertFinite(value: number, name: string): void {
  if (!Number.isFinite(value)) {
    throw new TypeError(`trackConstraints: ${name} must be a finite number, got ${value}`)
  }
}

function round3(t: number): number {
  return Math.round(t * 1000) / 1000
}

/** Snap `time` to the nearest multiple of `step` (bit-identical to Math.round(t*100)/100 for 0.01). */
export function snapToStep(time: number, step: number = SNAP_STEP): number {
  assertFinite(time, "time")
  if (step <= 0) throw new TypeError("trackConstraints: step must be positive")
  const inv = 1 / step
  return Math.round(time * inv) / inv
}

// ------------------------------------------------------------------
// M1-1: neighbor bounds + main-track cue constraint
// ------------------------------------------------------------------

export interface TrackNeighborBounds {
  /** End of the previous segment on the same track (null when none). */
  prevEnd: number | null
  /** Start of the next segment on the same track (null when none). */
  nextStart: number | null
}

type GeomSegment = Pick<Segment, "id" | "start" | "end">

/**
 * Neighbor bounds of `segmentId` on its own track. Segments whose id is in
 * `movedIds` are exempt (multi-select / linkage drags must not bound each
 * other). Tolerates unsorted input (defensive sort, O(n log n)).
 */
export function getTrackNeighborBounds(
  segments: ReadonlyArray<GeomSegment>,
  segmentId: string,
  movedIds?: ReadonlySet<string>,
): TrackNeighborBounds {
  const sorted = [...segments].sort((a, b) => a.start - b.start)
  const idx = sorted.findIndex(s => s.id === segmentId)
  if (idx === -1) return { prevEnd: null, nextStart: null }
  let prevEnd: number | null = null
  for (let i = idx - 1; i >= 0; i--) {
    if (movedIds?.has(sorted[i].id)) continue
    prevEnd = sorted[i].end
    break
  }
  let nextStart: number | null = null
  for (let i = idx + 1; i < sorted.length; i++) {
    if (movedIds?.has(sorted[i].id)) continue
    nextStart = sorted[i].start
    break
  }
  return { prevEnd, nextStart }
}

export type ConstrainCueResult =
  | { ok: true; start: number; end: number }
  | { ok: false; reason: "gap-too-narrow"; gap: number }

/**
 * Main-track "clamp into the neighbor gap" rule:
 *
 * 1. Clamp [start, end] into (prevEnd, nextStart); +-Infinity bounds from
 *    missing neighbors are naturally no-ops.
 * 2. If the gap itself is narrower than `minDuration` -> blocked (caller
 *    must reject the move).
 * 3. If the clamped range got narrower than `minDuration` while the
 *    original width was sufficient -> slide inside the gap keeping the
 *    ORIGINAL width: hug the previous segment; if that overflows hug the
 *    next; if that overflows too (original width > gap) cap to the gap.
 *
 * Only neighbor geometry lives here; global [0, duration] clamping stays
 * with the caller (clampTime).
 */
export function constrainCueRangeToTrack(
  start: number,
  end: number,
  bounds: TrackNeighborBounds,
  minDuration: number = MIN_SEGMENT_DURATION,
): ConstrainCueResult {
  assertFinite(start, "start")
  assertFinite(end, "end")
  if (end < start) [start, end] = [end, start]
  const originalWidth = end - start

  const lo = bounds.prevEnd ?? -Infinity
  const hi = bounds.nextStart ?? Infinity
  if (hi - lo < minDuration - EPSILON) {
    return { ok: false, reason: "gap-too-narrow", gap: Number.isFinite(hi - lo) ? hi - lo : 0 }
  }

  let s = Math.max(start, lo)
  let e = Math.min(end, hi)
  if (e - s < minDuration - EPSILON && originalWidth >= minDuration - EPSILON) {
    // Keep the ORIGINAL width while sliding. Note: with dur = min(width, gap)
    // the hug-next branch would be unreachable (hug-prev overflows iff
    // dur > gap); keeping the raw width keeps the fallback chain live.
    const dur = originalWidth
    s = lo
    e = lo + dur
    if (e > hi + EPSILON) {
      e = hi
      s = hi - dur
      if (s < lo - EPSILON) {
        s = lo
      }
    }
  }
  return { ok: true, start: round3(s), end: round3(e) }
}

// ------------------------------------------------------------------
// M1-2: extension-track constraints
// ------------------------------------------------------------------

/**
 * Global [0, duration] clamp + minimum duration + whole-millisecond
 * rounding. Degenerate media (duration <= min) collapses to [0, round3(duration)].
 */
export function clampExtensionRange(
  start: number,
  end: number,
  duration: number,
  minDuration: number = MIN_SEGMENT_DURATION,
): { start: number; end: number } {
  assertFinite(start, "start")
  assertFinite(end, "end")
  assertFinite(duration, "duration")
  if (duration <= 0) return { start: 0, end: 0 }
  if (duration <= minDuration) return { start: 0, end: round3(duration) }

  let s = Math.min(Math.max(0, start), duration - minDuration)
  let e = Math.min(Math.max(end, minDuration), duration)
  if (e - s < minDuration) {
    if (s + minDuration <= duration) {
      e = s + minDuration
    } else {
      s = duration - minDuration
      e = duration
    }
  }
  return { start: round3(s), end: round3(e) }
}

/**
 * O(n) overlap probe. Touching edges (gap <= epsilon) do NOT count as
 * overlap; self and moved segments are skipped. Extension tracks forbid
 * overlap outright -- callers must reject the move when this returns true
 * (no slide-in-place on extension lanes).
 */
export function extensionRangeOverlapsNeighbors(
  start: number,
  end: number,
  segments: ReadonlyArray<GeomSegment>,
  segmentId: string,
  movedIds?: ReadonlySet<string>,
  epsilon: number = EPSILON,
): boolean {
  assertFinite(start, "start")
  assertFinite(end, "end")
  for (const s of segments) {
    if (s.id === segmentId) continue
    if (movedIds?.has(s.id)) continue
    if (start < s.end - epsilon && end > s.start + epsilon) return true
  }
  return false
}

// ------------------------------------------------------------------
// M1-3: linkage follow + reconcile (semantic twin lives in the backend)
// ------------------------------------------------------------------

export interface ReconcileCounters {
  /** Segment kept but compressed to the uncovered side. */
  squeezed: number
  /** Segment deleted (uncovered side below min duration, or fully covered). */
  removed: number
  /** Bindings dissolved because their extension segment was removed (1:1 model: == removed). */
  unbound: number
}

export interface ReconcileResult {
  /** Surviving extension segments with their new geometry. */
  segments: Array<Pick<Segment, "id" | "start" | "end">>
  /** Extension segment ids removed by reconcile. Callers derive unbound bindings from this (1:1). */
  removedIds: string[]
  counters: ReconcileCounters
}

/**
 * Passive-side resolution after the main track moved: every extension
 * segment intersecting `covered` keeps its longest uncovered side; if that
 * side is shorter than `minDuration` the segment is deleted. `covered` is
 * READ-ONLY input -- reconcile never rewrites the main track (red line).
 *
 * Note (SPEC errata M1-3): the original signature carried
 * `unboundBindingIds` in the result; bindings are not known to this
 * function, so dissolution is derived by the caller from `removedIds`
 * (1:1 model: removed segment == unbound binding).
 */
export function reconcileExtensionTrack(
  extSegments: ReadonlyArray<GeomSegment>,
  covered: ReadonlyArray<{ start: number; end: number }>,
  minDuration: number = MIN_SEGMENT_DURATION,
): ReconcileResult {
  const counters: ReconcileCounters = { squeezed: 0, removed: 0, unbound: 0 }
  const out: Array<Pick<Segment, "id" | "start" | "end">> = []
  const removedIds: string[] = []

  for (const seg of extSegments) {
    // Collect uncovered sub-ranges of [seg.start, seg.end].
    const gaps: Array<{ start: number; end: number }> = []
    let cursor = seg.start
    for (const c of covered) {
      if (c.end <= cursor + EPSILON || c.start >= seg.end - EPSILON) continue
      if (c.start > cursor + EPSILON) {
        gaps.push({ start: cursor, end: Math.min(c.start, seg.end) })
      }
      cursor = Math.max(cursor, Math.min(c.end, seg.end))
      if (cursor >= seg.end - EPSILON) break
    }
    if (cursor < seg.end - EPSILON) {
      gaps.push({ start: cursor, end: seg.end })
    }

    if (gaps.length === 0) {
      // Fully covered -> delete.
      removedIds.push(seg.id)
      counters.removed++
      counters.unbound++
      continue
    }

    // Longest uncovered side wins.
    const best = gaps.reduce((a, b) => (b.end - b.start > a.end - a.start ? b : a))
    if (best.end - best.start >= minDuration - EPSILON) {
      if (
        Math.abs(best.start - seg.start) > EPSILON ||
        Math.abs(best.end - seg.end) > EPSILON
      ) {
        counters.squeezed++
      }
      out.push({ id: seg.id, start: round3(best.start), end: round3(best.end) })
    } else {
      removedIds.push(seg.id)
      counters.removed++
      counters.unbound++
    }
  }

  return { segments: out, removedIds, counters }
}

/**
 * Main -> extension delta follow: whole-span shift for moves (equal edge
 * deltas), per-edge follow for trims (both edges stack). The output is a
 * CANDIDATE geometry -- callers must still run
 * `extensionRangeOverlapsNeighbors` / `reconcileExtensionTrack`.
 */
export function syncBoundExtensionForMain(
  mainBefore: Pick<Segment, "start" | "end">,
  mainAfter: Pick<Segment, "start" | "end">,
  ext: Pick<Segment, "start" | "end">,
): { start: number; end: number } {
  const dStart = mainAfter.start - mainBefore.start
  const dEnd = mainAfter.end - mainBefore.end
  return { start: ext.start + dStart, end: ext.end + dEnd }
}

/**
 * Wholescale offset rebuild: offset = ext - main, rounded to the
 * millisecond. The ONLY sanctioned way to produce binding offsets -- no
 * incremental maintenance anywhere (red line M0-3).
 */
export function rebuildBindingOffsets(
  main: Pick<Segment, "start" | "end">,
  ext: Pick<Segment, "start" | "end">,
): Pick<TrackBinding, "start_offset" | "end_offset"> {
  return {
    start_offset: round3(ext.start - main.start),
    end_offset: round3(ext.end - main.end),
  }
}

// ------------------------------------------------------------------
// M1-4: extension -> main reverse constraint (ported, UI not wired this
// release -- dragging an extension segment only moves the extension
// segment; see SPEC M1-4 ruling)
// ------------------------------------------------------------------

export interface BoundPanelEditResult {
  ok: boolean
  mainStart: number
  mainEnd: number
  /** Actually applied shift (<= |delta| after neighbor clamping). */
  shifted: number
}

/**
 * Reverse mapping: a drag of `delta` on a bound extension segment proposes
 * moving the main segment by the same amount; the proposal is constrained
 * by the main track's neighbor gap and mapped back. `ok=false` means the
 * gap cannot host the segment at all (caller keeps the original geometry).
 */
export function constrainBoundExtensionPanelEdit(
  delta: number,
  main: Pick<Segment, "start" | "end">,
  bounds: TrackNeighborBounds,
  minDuration: number = MIN_SEGMENT_DURATION,
): BoundPanelEditResult {
  assertFinite(delta, "delta")
  const proposedStart = main.start + delta
  const proposedEnd = main.end + delta
  const r = constrainCueRangeToTrack(proposedStart, proposedEnd, bounds, minDuration)
  if (!r.ok) {
    return { ok: false, mainStart: main.start, mainEnd: main.end, shifted: 0 }
  }
  return { ok: true, mainStart: r.start, mainEnd: r.end, shifted: r.start - main.start }
}
